drivers/counter/counter_mcux_pit.c

```diff
@@ -211,48 +212,43 @@ static const struct counter_driver_api mcux_pit_driver_api = {
 	.get_top_value = mcux_pit_get_top_value,
 };
 
-/*
- * This driver is single-instance. If the devicetree contains multiple
- * instances, this will fail and the driver needs to be revisited.
- */
-BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) <= 1,
-	     "unsupported pit instance");
-
-static struct mcux_pit_data mcux_pit_data_0;
-
-static void mcux_pit_irq_config_0(const struct device *dev);
-
-static const struct mcux_pit_config mcux_pit_config_0 = {
-	.info = {
-		.max_top_value = UINT32_MAX,
-		.channels = 1,
-		.freq = DT_INST_PROP(0, clock_frequency),
-	},
-	.base = (PIT_Type *)DT_INST_REG_ADDR(0),
-	.pit_channel = DT_INST_PROP(0, pit_channel),
-	.irq_config_func = mcux_pit_irq_config_0,
-};
-
-DEVICE_DT_INST_DEFINE(0, &mcux_pit_init, NULL,
-		    &mcux_pit_data_0, &mcux_pit_config_0, POST_KERNEL,
-		    CONFIG_COUNTER_INIT_PRIORITY, &mcux_pit_driver_api);
+#define COUNTER_MCUX_PIT_DEVICE(n)						\
+	static void mcux_pit_irq_config_##n(const struct device *dev);		\
+	static struct mcux_pit_data mcux_pit_data_##n;				\
+	static const struct mcux_pit_config mcux_pit_config_##n = {		\
+		.info = {							\
+			.max_top_value = UINT32_MAX,				\
+			.channels = 1,						\
+			.freq = DT_INST_PROP(n, clock_frequency),		\
+		},								\
+		.base = (PIT_Type *)DT_INST_REG_ADDR(n),			\
+		.pit_channel = DT_INST_PROP(n, pit_channel),			\
+		.pit_period = DT_INST_PROP(n, pit_period),			\
+		.irq_config_func = mcux_pit_irq_config_##n,			\
+	};									\
+										\
+	DEVICE_DT_INST_DEFINE(n, &mcux_pit_init, NULL,				\
+			&mcux_pit_data_##n, &mcux_pit_config_##n, POST_KERNEL,	\
+			CONFIG_COUNTER_INIT_PRIORITY, &mcux_pit_driver_api);	\
+										\
+	static void mcux_pit_irq_config_##n(const struct device *dev)		\
+	{									\
+		IRQ_CONNECT(DT_INST_IRQ_BY_IDX(0, 0, irq),			\
+			DT_INST_IRQ_BY_IDX(0, 0, priority), mcux_pit_isr,	\
+			DEVICE_DT_INST_GET(0), 0);				\
+		irq_enable(DT_INST_IRQ_BY_IDX(0, 0, irq));			\
+		IRQ_CONNECT(DT_INST_IRQ_BY_IDX(0, 1, irq),			\
+			DT_INST_IRQ_BY_IDX(0, 1, priority), mcux_pit_isr,	\
+			DEVICE_DT_INST_GET(0), 0);				\
+		irq_enable(DT_INST_IRQ_BY_IDX(0, 1, irq));			\
+		IRQ_CONNECT(DT_INST_IRQ_BY_IDX(0, 2, irq),			\
+			DT_INST_IRQ_BY_IDX(0, 2, priority), mcux_pit_isr,	\
+			DEVICE_DT_INST_GET(0), 0);				\
+		irq_enable(DT_INST_IRQ_BY_IDX(0, 2, irq));			\
+		IRQ_CONNECT(DT_INST_IRQ_BY_IDX(0, 3, irq),			\
+			DT_INST_IRQ_BY_IDX(0, 3, priority), mcux_pit_isr,	\
+			DEVICE_DT_INST_GET(0), 0);				\
+		irq_enable(DT_INST_IRQ_BY_IDX(0, 3, irq));			\
```

Comment:
While we are at improving this driver, can we do something of this static dependence on the number of interrupts. 
The PR to edma driver has a method where it finds the number of interrupts from the device tree property, we could do the same here.
https://github.com/zephyrproject-rtos/zephyr/pull/61311/commits/cf00e46c97c75f3ddcb64c641af91c6845b0bb65

Reply:
Is done in https://github.com/zephyrproject-rtos/zephyr/commit/928e948f7b3810ebb0fa703b2011b10436889a74 😀